File: core/src/main/java/com/pmcl/core/mods/ModDropInstaller.java
package com.pmcl.core.mods;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.pmcl.core.LauncherConfig;
import com.pmcl.core.market.ModrinthClient;
import com.pmcl.core.preferences.Preferences;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 拖放安装器：解析拖入的 mod jar + SHA1 反查 Modrinth + 拷贝到目标 mods 目录。
 * <p>
 * 流程：
 * <ol>
 *   <li>{@link #analyze(List)}：对每个 jar 调用 {@link ModScanner#parseJar(Path)} 拿元数据，
 *       计算 SHA1，批量调用 {@link ModrinthClient#batchCheckBySha1(List)} 反查
 *       game_versions / loaders，返回 {@link ModDropInfo} 列表</li>
 *   <li>{@link #installTo(ModDropInfo, String, String)}：拷贝 jar 到目标 mods 目录
 *       （版本隔离 → instances/&lt;versionId&gt;/mods/，否则 mods/&lt;gameVersion&gt;/）</li>
 * </ol>
 * <p>
 * 设计：所有 IO 在调用方线程执行（同步方法），让 UI 决定是否切到 IO 协程。
 * 网络失败时不抛异常，降级为 modrinthFound=false，UI 仍允许用户手动选择版本。
 */
public final class ModDropInstaller {

    private final LauncherConfig config;
    private final ModrinthClient modrinth;

    public ModDropInstaller(LauncherConfig config, ModrinthClient modrinth) {
        this.config = config;
        this.modrinth = modrinth;
    }

    /**
     * 批量解析拖入的 jar 文件。
     * <p>
     * 对每个 jar：
     * <ol>
     *   <li>ModScanner.parseJar 拿 modId/name/version/loader</li>
     *   <li>计算文件 SHA1</li>
     * </ol>
     * 然后一次性批量调用 Modrinth batchCheckBySha1 反查所有 jar 的兼容版本信息。
     *
     * @param jarPaths 拖入的 jar 文件路径列表
     * @return 解析结果列表（顺序与输入一致），解析失败的 jar 也会返回带 parseError 的项
     */
    public List<ModDropInfo> analyze(List<Path> jarPaths) {
        if (jarPaths == null || jarPaths.isEmpty()) return Collections.emptyList();

        // 第一阶段：解析 jar 元数据 + 计算 SHA1
        List<ModMeta> metas = new ArrayList<>(jarPaths.size());
        List<String> sha1s = new ArrayList<>(jarPaths.size());
        List<String> parseErrors = new ArrayList<>(jarPaths.size());
        for (Path p : jarPaths) {
            ModMeta meta;
            String err = null;
            try {
                meta = ModScanner.parseJar(p);
            } catch (Throwable t) {
                meta = null;
                err = t.getMessage() != null ? t.getMessage() : t.getClass().getSimpleName();
            }
            metas.add(meta);
            parseErrors.add(err);
            String sha1 = null;
            try {
                sha1 = computeSha1(p);
            } catch (IOException ignored) {
                // SHA1 计算失败：仍允许安装，只是无法反查 Modrinth
            }
            sha1s.add(sha1);
        }

        // 第二阶段：批量反查 Modrinth（仅对有 SHA1 的 jar）
        Map<String, JsonObject> modrinthMap = Collections.emptyMap();
        if (modrinth != null) {
            List<String> nonNullSha1s = new ArrayList<>();
            for (String s : sha1s) {
                if (s != null && !s.isEmpty() && !nonNullSha1s.contains(s)) {
                    nonNullSha1s.add(s);
                }
            }
            if (!nonNullSha1s.isEmpty()) {
                try {
                    modrinthMap = modrinth.batchCheckBySha1(nonNullSha1s);
                } catch (Throwable ignored) {
                    // 网络失败：降级为 modrinthFound=false
                    modrinthMap = Collections.emptyMap();
                }
            }
        }

        // 第三阶段：组装 ModDropInfo
        List<ModDropInfo> result = new ArrayList<>(jarPaths.size());
        for (int i = 0; i < jarPaths.size(); i++) {
            Path jarPath = jarPaths.get(i);
            ModMeta meta = metas.get(i);
            String sha1 = sha1s.get(i);
            String err = parseErrors.get(i);

            String modId, name, version, loader, authors, description;
            if (meta != null) {
                modId = meta.getModId();
                name = meta.getName();
                version = meta.getVersion();
                loader = meta.getLoader();
                authors = meta.getAuthors();
                description = meta.getDescription();
            } else {
                modId = "";
                name = jarPath.getFileName().toString();
                version = "";
                loader = "unknown";
                authors = "";
                description = "";
            }

            List<String> gameVersions = Collections.emptyList();
            List<String> loaders = Collections.emptyList();
            boolean found = false;
            if (sha1 != null && modrinthMap.containsKey(sha1)) {
                JsonObject v = modrinthMap.get(sha1);
                if (v != null) {
                    gameVersions = jsonArrToStrings(v, "game_versions");
                    loaders = jsonArrToStrings(v, "loaders");
                    found = !gameVersions.isEmpty() || !loaders.isEmpty();
                }
            }
            result.add(new ModDropInfo(modId, name, version, loader, authors, description,
                    jarPath, sha1, gameVersions, loaders, found, err));
        }
        return result;
    }

    /**
     * 把已解析的 mod jar 拷贝到目标 mods 目录。
     * <p>
     * 路径推导：
     * <ul>
     *   <li>版本隔离开启且 versionId 非空：{@code ~/.pmcl/instances/<versionId>/mods/}</li>
     *   <li>否则：{@code ~/.pmcl/mods/<gameVersion>/}（gameVersion 为空则直接 mods/）</li>
     * </ul>
     * 同名文件存在时覆盖（让用户能拖入新版 jar 更新）。
     *
     * @param info        已解析的 mod 信息
     * @param versionId   目标版本 ID（用于版本隔离），可为 null
     * @param gameVersion 目标 MC 版本号（如 "1.20.1"），可为 null
     * @return 拷贝目标路径
     * @throws IOException 拷贝失败
     */
    public Path installTo(ModDropInfo info, String versionId, String gameVersion) throws IOException {
        Path modsDir;
        if (preferences != null && preferences.isVersionIsolation()
                && versionId != null && !versionId.isEmpty()) {
            modsDir = config.getWorkDir().resolve("instances").resolve(versionId).resolve("mods");
        } else {
            modsDir = config.getWorkDir().resolve("mods");
            if (gameVersion != null && !gameVersion.isEmpty()) {
                modsDir = modsDir.resolve(gameVersion);
            }
        }
        Files.createDirectories(modsDir);
        Path target = modsDir.resolve(info.getJarPath().getFileName());
        Files.copy(info.getJarPath(), target, StandardCopyOption.REPLACE_EXISTING);
        return target;
    }

    /** 可选注入的 Preferences，用于判断版本隔离开关 */
    private Preferences preferences;

    public void setPreferences(Preferences prefs) {
        this.preferences = prefs;
    }

    // ==================== 辅助 ====================

    /** 计算文件 SHA1（hex 小写） */
    private static String computeSha1(Path path) throws IOException {
        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-1");
        } catch (Exception e) {
            return null;
        }
        try (InputStream in = Files.newInputStream(path)) {
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) {
                md.update(buf, 0, n);
            }
        }
        byte[] digest = md.digest();
        StringBuilder sb = new StringBuilder(digest.length * 2);
        for (byte b : digest) {
            sb.append(Character.forDigit((b >> 4) & 0xF, 16));
            sb.append(Character.forDigit(b & 0xF, 16));
        }
        return sb.toString();
    }

    private static List<String> jsonArrToStrings(JsonObject o, String key) {
        if (o == null || !o.has(key) || !o.get(key).isJsonArray()) return Collections.emptyList();
        JsonArray arr = o.getAsJsonArray(key);
        List<String> list = new ArrayList<>(arr.size());
        for (JsonElement e : arr) {
            if (e.isJsonPrimitive()) list.add(e.getAsString());
        }
        return list;
    }
}

File: pcmlAndroid/core/src/main/java/com/lash/pmcl/core/mods/ModDropInstaller.kt
package com.lash.pmcl.core.mods

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.lash.pmcl.core.market.ModrinthClient
import com.lash.pmcl.core.paths.PmclPaths
import com.lash.pmcl.core.preferences.Preferences
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.util.ArrayList

/**
 * 拖放安装器：解析拖入的 mod jar + SHA1 反查 Modrinth + 拷贝到目标 mods 目录。
 *
 * 流程：
 * 1. [analyze]：对每个 jar 调用 [ModScanner.parseJar] 拿元数据，
 *    计算 SHA1，批量调用 [ModrinthClient.batchCheckBySha1] 反查
 *    game_versions / loaders，返回 [ModDropInfo] 列表
 * 2. [installTo]：拷贝 jar 到目标 mods 目录
 *    （版本隔离 → instances/<versionId>/mods/，否则 mods/<gameVersion>/）
 *
 * 设计：所有 IO 在调用方线程执行（同步方法），让 UI 决定是否切到 IO 协程。
 * 网络失败时不抛异常，降级为 modrinthFound=false，UI 仍允许用户手动选择版本。
 */
class ModDropInstaller(
    private val paths: PmclPaths,
    @Suppress("unused") private val preferences: Preferences,
    private val modrinth: ModrinthClient
) {

    /**
     * 批量解析拖入的 jar 文件。
     *
     * 对每个 jar：
     * 1. ModScanner.parseJar 拿 modId/name/version/loader
     * 2. 计算文件 SHA1
     *
     * 然后一次性批量调用 Modrinth batchCheckBySha1 反查所有 jar 的兼容版本信息。
     *
     * @param jarPaths 拖入的 jar 文件路径列表
     * @return 解析结果列表（顺序与输入一致），解析失败的 jar 也会返回带 parseError 的项
     */
    fun analyze(jarPaths: List<Path>?): List<ModDropInfo> {
        if (jarPaths.isNullOrEmpty()) return emptyList()

        // 第一阶段：解析 jar 元数据 + 计算 SHA1
        val metas = ArrayList<ModMeta?>(jarPaths.size)
        val sha1s = ArrayList<String?>(jarPaths.size)
        val parseErrors = ArrayList<String?>(jarPaths.size)
        for (p in jarPaths) {
            var meta: ModMeta? = null
            var err: String? = null
            try {
                meta = ModScanner.parseJar(p)
            } catch (t: Throwable) {
                err = t.message ?: t.javaClass.simpleName
            }
            metas.add(meta)
            parseErrors.add(err)
            var sha1: String? = null
            try {
                sha1 = computeSha1(p)
            } catch (_: IOException) {
                // SHA1 计算失败：仍允许安装，只是无法反查 Modrinth
            }
            sha1s.add(sha1)
        }

        // 第二阶段：批量反查 Modrinth（仅对有 SHA1 的 jar）
        var modrinthMap: Map<String, JsonObject> = emptyMap()
        val nonNullSha1s = ArrayList<String>()
        for (s in sha1s) {
            if (!s.isNullOrEmpty() && !nonNullSha1s.contains(s)) {
                nonNullSha1s.add(s)
            }
        }
        if (nonNullSha1s.isNotEmpty()) {
            try {
                modrinthMap = modrinth.batchCheckBySha1(nonNullSha1s)
            } catch (_: Throwable) {
                // 网络失败：降级为 modrinthFound=false
                modrinthMap = emptyMap()
            }
        }

        // 第三阶段：组装 ModDropInfo
        val result = ArrayList<ModDropInfo>(jarPaths.size)
        for (i in jarPaths.indices) {
            val jarPath = jarPaths[i]
            val meta = metas[i]
            val sha1 = sha1s[i]
            val err = parseErrors[i]

            val modId: String
            val name: String
            val version: String
            val loader: String
            val authors: String
            val description: String
            if (meta != null) {
                modId = meta.modId
                name = meta.name
                version = meta.version
                loader = meta.loader
                authors = meta.authors
                description = meta.description
            } else {
                modId = ""
                name = jarPath.fileName.toString()
                version = ""
                loader = "unknown"
                authors = ""
                description = ""
            }

            var gameVersions: List<String> = emptyList()
            var loaders: List<String> = emptyList()
            var found = false
            if (sha1 != null && modrinthMap.containsKey(sha1)) {
                val v = modrinthMap[sha1]
                if (v != null) {
                    gameVersions = jsonArrToStrings(v, "game_versions")
                    loaders = jsonArrToStrings(v, "loaders")
                    found = gameVersions.isNotEmpty() || loaders.isNotEmpty()
                }
            }
            result.add(
                ModDropInfo(
                    modId, name, version, loader, authors, description,
                    jarPath, sha1, gameVersions, loaders, found, err
                )
            )
        }
        return result
    }

    /**
     * 把已解析的 mod jar 拷贝到目标 mods 目录。
     *
     * 路径推导：
     * - versionId 非空（版本隔离）：instances/<versionId>/mods/
     * - 否则：mods/<gameVersion>/（gameVersion 为空则直接 mods/）
     *
     * 同名文件存在时覆盖（让用户能拖入新版 jar 更新）。
     *
     * @param info        已解析的 mod 信息
     * @param versionId   目标版本 ID（用于版本隔离），可为 null
     * @param gameVersion 目标 MC 版本号（如 "1.20.1"），可为 null
     * @return 拷贝目标路径
     * @throws IOException 拷贝失败
     */
    @Throws(IOException::class)
    fun installTo(info: ModDropInfo, versionId: String?, gameVersion: String?): Path {
        val modsDir: Path
        if (!versionId.isNullOrEmpty()) {
            // 版本隔离模式：instances/<versionId>/mods
            requireSafeName(versionId)
            val instancesRoot = paths.instances.toAbsolutePath().normalize()
            val instanceDir = instancesRoot.resolve(versionId).normalize()
            if (!instanceDir.startsWith(instancesRoot)) {
                throw IOException("versionId path escapes instances dir: $versionId")
            }
            modsDir = instanceDir.resolve("mods")
        } else {
            val baseModsDir = paths.minecraftWorkDir.resolve("mods")
            modsDir = if (!gameVersion.isNullOrEmpty()) {
                requireSafeName(gameVersion)
                val modsRoot = baseModsDir.toAbsolutePath().normalize()
                val resolved = modsRoot.resolve(gameVersion).normalize()
                if (!resolved.startsWith(modsRoot)) {
                    throw IOException("gameVersion path escapes mods dir: $gameVersion")
                }
                resolved
            } else {
                baseModsDir
            }
        }
        val modsAbs = modsDir.toAbsolutePath().normalize()
        Files.createDirectories(modsAbs)
        val jarName = info.jarPath.fileName
            ?: throw IOException("非法模组文件名")
        val fileName = jarName.toString()
        if (fileName.isBlank() || fileName.contains("..") || fileName.contains("/") ||
            fileName.contains("\\") || fileName.indexOf('\u0000') >= 0
        ) {
            throw IOException("非法模组文件名: $fileName")
        }
        val target = modsAbs.resolve(fileName).normalize()
        if (!target.startsWith(modsAbs)) {
            throw IOException("模组路径越界: $fileName")
        }
        Files.copy(info.jarPath, target, StandardCopyOption.REPLACE_EXISTING)
        return target
    }

    // ==================== 辅助 ====================

    /** 校验名称不含路径穿越字符（替代桌面版 InstanceManager.requireSafeInstanceId） */
    private fun requireSafeName(name: String) {
        if (name.contains("..") || name.contains("/") ||
            name.contains("\\") || name.indexOf('\u0000') >= 0
        ) {
            throw IOException("非法名称: $name")
        }
    }

    /** 计算文件 SHA1（hex 小写） */
    @Throws(IOException::class)
    private fun computeSha1(path: Path): String? {
        val md: MessageDigest = try {
            MessageDigest.getInstance("SHA-1")
        } catch (e: Exception) {
            return null
        }
        Files.newInputStream(path).use { inp ->
            val buf = ByteArray(8192)
            var n: Int
            while (inp.read(buf).also { n = it } > 0) {
                md.update(buf, 0, n)
            }
        }
        val digest = md.digest()
        val sb = StringBuilder(digest.size * 2)
        for (b in digest) {
            sb.append(Character.forDigit((b.toInt() shr 4) and 0xF, 16))
            sb.append(Character.forDigit(b.toInt() and 0xF, 16))
        }
        return sb.toString()
    }

    private fun jsonArrToStrings(o: JsonObject?, key: String): List<String> {
        if (o == null || !o.has(key) || !o.get(key).isJsonArray) return emptyList()
        val arr: JsonArray = o.getAsJsonArray(key)
        val list = ArrayList<String>(arr.size())
        for (e: JsonElement in arr) {
            if (e.isJsonPrimitive) list.add(e.asString)
        }
        return list
    }
}
